R/TealAppDriver.R

```diff
@@ -554,18 +553,77 @@ TealAppDriver <- R6::R6Class( # nolint: object_name.
       filter_panel = character(0)
     ),
     # private methods ----
-    set_active_ns = function() {
+    # Helper function to extract wrapper ID from selector and take first match if multiple found
+    extract_wrapper_id = function(selector) {
+      wrapper_id <- sub(
+        "^#",
+        "",
+        self$get_attr(selector = selector, attribute = "href")
+      )
+      # Take first match if multiple found
+      if (length(wrapper_id) > 1) {
+        wrapper_id <- wrapper_id[1]
+      }
+      wrapper_id
+    },
+    # Helper function to check if wrapper ID is valid
+    is_valid_wrapper_id = function(wrapper_id) {
+      length(wrapper_id) == 1 && wrapper_id != "" && !is.na(wrapper_id)
+    },
+    set_active_ns = function(sleep_time = 0.5) {
+      # Although wait_for_idle() is called before set_active_ns(), it only ensures Shiny is not processing.
+      # wait_for_page_stability() is needed here to ensure the DOM/UI is fully rendered and stable
+      # before trying to extract the namespace.
+      private$wait_for_page_stability()
+
       all_inputs <- self$get_values()$input
       active_tab_inputs <- all_inputs[grepl("-active_module_id$", names(all_inputs))]
 
-      active_wrapper_id <- sub(
-        "^#",
-        "",
-        self$get_attr(
-          selector = sprintf(".teal-modules-tree li a.module-button[data-value='%s']", active_tab_inputs),
-          attribute = "href"
+      # If no active_module_id input found, find the selected/active tab button directly
+      if (!length(active_tab_inputs) || active_tab_inputs == "") {
+        active_wrapper_id <- private$extract_wrapper_id(
+          ".teal-modules-tree li a.module-button.active, .teal-modules-tree li a.module-button[aria-selected='true']"
         )
-      )
+        # If still not found, try any module button with a wrapper href
+        if (!private$is_valid_wrapper_id(active_wrapper_id)) {
+          active_wrapper_id <- private$extract_wrapper_id(
+            ".teal-modules-tree li a.module-button[href*='-wrapper']:not([href='#'])"
+          )
+        }
+      } else {
+        active_wrapper_id <- private$extract_wrapper_id(
+          sprintf(".teal-modules-tree li a.module-button[data-value='%s']", active_tab_inputs)
+        )
+      }
+
+      # Ensure we have a valid wrapper ID
+      # get_attr returns character(0) when no elements found, or NA_character_ for missing attributes
+      if (!private$is_valid_wrapper_id(active_wrapper_id)) {
+        # Try one more time after a short wait - the page might still be loading
+        Sys.sleep(sleep_time)
+        active_wrapper_id <- private$extract_wrapper_id(
+          ".teal-modules-tree li a.module-button[href*='-wrapper']:not([href='#'])"
+        )
```

Comment:
It seems to make more sense to sleep and call the method again with `sleep_time = NULL` that skips this part if it fails